test/e2e/lib/lifecycle.go

```diff
@@ -0,0 +1,243 @@
+	Tracker    *lib.Tracker
+}
+
+var setStackDriverConfigOnce = sync.Once{}
+
+func (c *Client) SetupStackDriverMetrics(t *testing.T) {
+	setStackDriverConfigOnce.Do(func() {
+		err := c.Core.Kube.UpdateConfigMap("cloud-run-events", "config-observability", map[string]string{
+			"metrics.allow-stackdriver-custom-metrics":     "true",
+			"metrics.backend-destination":                  "stackdriver",
+			"metrics.stackdriver-custom-metrics-subdomain": "cloud.google.com",
+			"metrics.reporting-period-seconds":             "60",
+		})
+		if err != nil {
+			t.Fatalf("Unable to set the ConfigMap: %v", err)
+		}
+	})
+}
+
+// DuplicateSecret duplicates a secret from a namespace to a new namespace.
+func (c *Client) DuplicateSecret(t *testing.T, name, namespace string) {
+	cc := c.Core
+	secret, err := cc.Kube.Kube.CoreV1().Secrets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to find Secret: %q in Namespace: %q: %s", name, namespace, err)
+		return
+	}
+	newSecret := &corev1.Secret{}
+	newSecret.Name = name
+	newSecret.Namespace = c.Namespace
+	newSecret.Data = secret.Data
+	newSecret.StringData = secret.StringData
+	newSecret.Type = secret.Type
+	newSecret, err = cc.Kube.Kube.CoreV1().Secrets(c.Namespace).Create(newSecret)
+	if err != nil {
+		t.Fatalf("Failed to create Secret: %s; %v", c.Namespace, err)
+	}
+}
+
+const (
+	interval = 1 * time.Second
+	timeout  = 5 * time.Minute
+)
+
+// TODO(chizhg): move this function to knative/pkg/test
+// WaitForResourceReady waits until the specified resource in the given namespace are ready.
+func (c *Client) WaitUntilJobDone(namespace, name string) (string, error) {
+	cc := c.Core
+	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		job, err := cc.Kube.Kube.BatchV1().Jobs(namespace).Get(name, metav1.GetOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Println(namespace, name, "not found", err)
+				// keep polling
+				return false, nil
+			}
+			return false, err
+		}
+		return operations.IsJobComplete(job), nil
+	})
+	if err != nil {
+		return "", err
+	}
+
+	// poll until the pod is terminated.
+	err = wait.PollImmediate(interval, timeout, func() (bool, error) {
+		pod, err := operations.GetJobPodByJobName(context.TODO(), cc.Kube.Kube, namespace, name)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Println(namespace, name, "not found", err)
+				// keep polling
+				return false, nil
+			}
+			return false, err
+		}
+		if pod != nil {
+			for _, cs := range pod.Status.ContainerStatuses {
+				if cs.State.Terminated != nil {
+					return true, nil
+				}
+			}
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return "", err
+	}
+	pod, err := operations.GetJobPodByJobName(context.TODO(), cc.Kube.Kube, namespace, name)
+	if err != nil {
+		return "", err
+	}
+	return operations.GetFirstTerminationMessage(pod), nil
+}
+
+func (c *Client) LogsFor(namespace, name string, tm *metav1.TypeMeta) (string, error) {
```

Comment:
I've seen this somewhere else, worth moving to knative/pkg/test as well? maybe add a TODO as with the previous one?

Reply:
Added a TODO comment.